From: Hannes Reinecke <hare@suse.de>
Date: Wed, 14 Mar 2018 13:53:48 +0100
Subject: [PATCH] blk-mq: stop 'delayed_run_work' in blk_mq_stop_hw_queue()
Patch-Mainline: Never, SLE12-SP3 specific
References: bsc#1084967

The hardware context has three different workqueues, and we need
to stop all of them when calling blk_mq_stop_hw_queue().
Otherwise the workqueue will be running on a stale blk_mq_hw_ctx
structure and cause a kernel oops.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 block/blk-mq.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index b85ce94a61d7..992b93021f03 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -1091,6 +1091,7 @@ void blk_mq_stop_hw_queue(struct blk_mq_hw_ctx *hctx)
 {
 	cancel_work(&hctx->run_work);
 	cancel_delayed_work(&hctx->delay_work);
+	cancel_delayed_work(&hctx->delayed_run_work);
 	set_bit(BLK_MQ_S_STOPPED, &hctx->state);
 }
 EXPORT_SYMBOL(blk_mq_stop_hw_queue);
-- 
2.12.3

